optaweb-employee-rostering-frontend/src/ui/pages/shift/ExportScheduleModal.tsx

```diff
@@ -121,10 +122,10 @@ export const ExportScheduleModal: React.FC<StateProps & OwnProps> = (props) => {
           <MultiTypeaheadSelectInput
             aria-label={t('forSpots')}
             emptyText={t('selectSpots')}
-            value={exportedSpots}
-            options={props.spotList}
+            value={exportedSpots.toArray()}
+            options={props.spotList.toArray()}
             optionToStringMap={spot => spot.name}
-            onChange={setExportedSpots}
+            onChange={newExportedSpots => setExportedSpots(List(newExportedSpots))}
```

Comment:
This seems like an unnecessary complexity. Notice that every time user changes the input and the component is rerendered, a conversion from array to List and from List to array happens. For what benefit?

I think we should either A) draw a clear line between the Redux store and the React UI and only use Immutable.js in the store, where side effects must be avoided or B) use Immutable collections everywhere and only do the `toArray()` conversion when sending data to third-party components.

I'm slightly inclined toward option A at this moment.

Reply:
I prefer A) since then we don't need to use a type cast or a type guard, although mutable values in components can cause hard to find bugs in the frontend (I am looking at you, moment...). Arrays aren't typically modified directly in components, so they less prone to bugs then dates.

Reply:
Thoughts on the selectors? (that is, should selectors return the immutable List, or the List.toArray? If the second, should we memoize the List or List.toArray? (the problem with memozing List.toArray is that an modifications to it in one component affects all components.)

Reply:
Ad Moment: #532.

Reply:
I would say let's do the minimum necessary changes to get rid of the in-house immutable collections support (`withElements` etc.) without introducing significant performance regression.

When this is done, we can take a look at the next steps:
- Replace `mapStateToProps` with [useSelector](https://react-redux.js.org/api/hooks)?
- [Memoize selectors with Reselect](https://programmerden.com/2020/04/06/optimize-your-redux-selectors-with-useselector-hook-and-memoize-them-with-reselect/) for performance optimization or not?

Reply:
Some `mapStateToProps` are used in class components, so we won't be able to use `useSelector` there (most new functional components already use `useSelector`.

The result is already memoized manually.